### PR TITLE
Task4

### PR DESCRIPTION
**Pull request by LeshaBleschik**

### CODE REVIEW

File: src/api/apiRoutes.js [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+
+// GAME PRODUCTS LIST
+router.get(`/products`, (req, res) => {
+  const listOfGames = getGames()
+  let searchQuery = req.query.search?.toLowerCase()
+  const categoryQuery = req.query.category
+  const filteredByCategory = listOfGames.filter((game) =>
+    game.category.includes(categoryQuery)
+  )
+  if (!searchQuery && !categoryQuery) {
```
Comment (GoodPassiveMan) [inline src/api/apiRoutes.js]: Why do you check query parameters right after filtering? Definitely need to do it before filtering

File: src/api/apiRoutes.js [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+    return res.send(listOfGames)
+  }
+  if (searchQuery === undefined) {
```
Comment (GoodPassiveMan) [inline src/api/apiRoutes.js]: Could be done at the line 17 without using `let` and with `||` or `??`

File: src/api/apiRoutes.js [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+    searchQuery = ""
+  }
+  if (searchQuery && !categoryQuery) {
```
Comment (GoodPassiveMan) [inline src/api/apiRoutes.js:25]: Why do you need several almost simalar filters and if-statements if you could use one single array of games and re-filter it when you need it.

Example
`
 let games = ...
 if (searchQuery) {
   games = games.filter(...)
 }
 
 if (categoryQuery) {
   games = games.filter(...)
 }
 
 res.send(games)
`

File: src/api/helpers.js [hunk 4]
```diff
@@ -0,0 +1,203 @@
...
+    description:
+      "Mauris suscipit dui eu elit pellentesque ornare. Etiam arcu quam, pulvinar ac lorem eget, placerat semper arcu. Suspendisse maximus dui non convallis faucibus. Nam enim erat, aliquam at iaculis consequat, consequat eget purus.",
+    ageAllowed: `${randomInteger(8, 18)}+`,
+    releaseDate: randomInteger(1350000000000, 1400000000000),
+  }
+  b.title = titles.title[i]
+  dataG.push(b)
+}
+
+export const gamesData = () => writeFile("games.json", dataG)
```
Comment (GoodPassiveMan) [inline src/api/helpers.js]: Why do you need this code for pasting games into json file? It should be done once and never reused again

File: src/pages/product-page/ProductsPage.tsx [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+import "./product-page.scss"
+import GameCard from "components/game-card/GameCard"
+import { Game } from "types"
+
+const ProductsPage = () => {
+  const [searchGames, setSearchGames] = useState<Game[]>([])
+
+  const location = useLocation()
+  const search = new URLSearchParams(location.search)
+  let category = search.get("category")
```
Comment (GoodPassiveMan) [inline src/pages/product-page/ProductsPage.tsx]: It is not crucial, but const could be used here. Shows that you'll never change that variable later

File: src/pages/products/pc/Pc.tsx [hunk 6]
```diff
@@ -0,0 +1,5 @@
+import React from "react"
+
+const Pc = () => <h1>PC</h1>
+
+export default Pc
```
Comment (GoodPassiveMan) [inline src/pages/products/pc/Pc.tsx]: I think those pc/ps/xbox components could be removed

File: src/pages/products/playstation/Playstation.tsx [hunk 7]
```diff
@@ -0,0 +1,5 @@
+import React from "react"
+
+const Playstation = () => <h1>Playstation 5</h1>
+
+export default Playstation
```
Comment (GoodPassiveMan) [inline src/pages/products/playstation/Playstation.tsx]: I think those pc/ps/xbox components could be removed

File: src/pages/products/xbox/Xbox.tsx [hunk 8]
```diff
@@ -0,0 +1,5 @@
+import React from "react"
+
+const Xbox = () => <h1>XBox One</h1>
+
+export default Xbox
```
Comment (GoodPassiveMan) [inline src/pages/products/xbox/Xbox.tsx]: I think those pc/ps/xbox components could be removed

File: src/routes.ts [hunk 9]
```diff
@@ -1,8 +1,11 @@
...
 export const SIGN_UP_PAGE = "/sign-up"
-export const PRODUCT_GAMES_PAGE = "/product/games"
-export const PRODUCT_CONSOLES_PAGE = "/product/consoles"
-export const PRODUCT_ACCESSORIES_PAGE = "/product/game-accessories"
+export const PRODUCTS_PAGE = "/products"
 export const WRONG_PATH = "*"
+
+// API REQUESTS
+export const GET_TOP_PRODUCTS = "http://localhost:5000/api/get-top-products"
+export const PRODUCTS_SEARCH = "http://localhost:5000/api/products"
```
Comment (GoodPassiveMan) [inline src/routes.ts]: Create separate file for api constants. This file is used for routing configs.

**Review by GoodPassiveMan — CHANGES_REQUESTED** [review]
Have left few comments to fix or discuss

File: src/pages/product-page/ProductsPage.tsx (hunk 5 above)
Comment (LeshaBleschik) [reply src/pages/product-page/ProductsPage.tsx]: Down to the file i re-assign the value of category. It's not very important but i thought it wood look slightly better

File: src/api/apiRoutes.js (hunk 1 above)
Comment (LeshaBleschik) [reply src/api/apiRoutes.js]: I use this filterArray only in places where i definitely know i pass certain condition. Following checking params does not affect this array by any means

File: src/pages/product-page/ProductsPage.tsx (hunk 5 above)
Comment (GoodPassiveMan) [reply src/pages/product-page/ProductsPage.tsx]: Okay, haven't seen that. 

Could you create a helper called `capitalize`.

`const capitalize = (value:  string): string => {
  if (!value.length) return value;
  return '${value[0].toUpperCase()}${value.slice(1)}'
}`

Usage:

`const category = capitalize(search.get('category') || '')`

Try not to use variable mutations if possible.

File: src/pages/product-page/ProductsPage.tsx [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+      .get(`${PRODUCTS_SEARCH}${location.search}`)
+      .then((response) => {
+        setSearchGames(response.data)
+      })
+      .catch((error) => {
+        console.log(error)
+      })
+  }, [location.search])
+
+  const memoizedCategory = useMemo(() => {
```
Comment (GoodPassiveMan) [inline src/pages/product-page/ProductsPage.tsx]: If you use `capitalize` then you wont need it

File: src/api/apiRoutes.js [hunk 11]
```diff
@@ -0,0 +1,36 @@
...
+    .slice(0, 3)
+  return res.send(gamesSortedByDate)
+})
+
+// GAME PRODUCTS LIST
+router.get(`/products`, (req, res) => {
+  let listOfGames = getGames()
+  const searchQuery = req.query.search?.toLowerCase() || ""
+  const categoryQuery = req.query.category
+  const filteredByCategory = listOfGames.filter((game) =>
```
Comment (GoodPassiveMan) [inline src/api/apiRoutes.js]: Are you sure that category is always provided?
I remember if you use search bar then you'll get `?search=bla-bla-bla` and category here isn't presented, so your request would look like `fetch('api/products?search=bla-bla-bla')` and you receive an undefined value for category. Please concider do an undefined check before doing filter

**Review by GoodPassiveMan — CHANGES_REQUESTED** [review]
2 small fixes requested

File: src/api/apiRoutes.js (hunk 11 above)
Comment (LeshaBleschik) [reply src/api/apiRoutes.js]: I did what you requested. I'm not arguing just want to share my thoughts on that part of code. We use filteredByCategory array only when we pass certain conditions like: ` if (categoryQuery && searchQuery)` or` if (categoryQuery)`. I got the point this array will be empty if  we have no any category, but in this case we don't use it at all. Is it like the thing to remove possible bugs and be more efficient?

**Review by GoodPassiveMan — APPROVED** [review]

**Pull request merged**